decorate_test.go

```diff
@@ -0,0 +1,463 @@
+		}
+
+		c := dig.New()
+		require.NoError(t, c.Provide(func() string { return "dog" }, dig.Group("animals")))
+		require.NoError(t, c.Provide(func() string { return "cat" }, dig.Group("animals")))
+		require.NoError(t, c.Provide(func() string { return "alpaca" }, dig.Group("animals")))
+
+		assert.NoError(t, c.Decorate(func(p Params) Result {
+			animals := p.Animals
+			for i := 0; i < len(animals); i++ {
+				animals[i] = "good " + animals[i]
+			}
+			return Result{
+				Animals: animals,
+			}
+		}))
+
+		assert.NoError(t, c.Invoke(func(p Params) {
+			assert.Contains(t, p.Animals, "good dog")
+		}))
+	})
+
+	t.Run("decorate with optional parameter", func(t *testing.T) {
+		c := dig.New()
+
+		type A struct{}
+		type Param struct {
+			dig.In
+
+			Values []string `group:"values"`
+			A      *A       `optional:"true"`
+		}
+
+		type Result struct {
+			dig.Out
+
+			Values []string `group:"values"`
+		}
+
+		require.NoError(t, c.Provide(func() string { return "a" }, dig.Group("values")))
+		require.NoError(t, c.Provide(func() string { return "b" }, dig.Group("values")))
+
+		require.NoError(t, c.Decorate(func(p Param) Result {
+			return Result{
+				Values: append(p.Values, "c"),
+			}
+		}))
+
+		assert.NoError(t, c.Invoke(func(p Param) {
+			assert.Equal(t, 3, len(p.Values))
+		}))
+	})
+
+	t.Run("replace a type completely", func(t *testing.T) {
+		t.Parallel()
+
+		c := dig.New()
+		type A struct {
+			From string
+		}
+
+		require.NoError(t, c.Provide(func() A {
+			assert.Fail(t, "provider shouldn't be called")
+			return A{From: "provider"}
+		}))
+		require.NoError(t, c.Decorate(func() A {
+			return A{From: "decorator"}
+		}))
+		require.NoError(t, c.Invoke(func(a A) {
+			assert.Equal(t, a.From, "decorator", "value should be from decorator")
+		}))
+	})
+
+	t.Run("group value decorator from parent and child", func(t *testing.T) {
+		t.Parallel()
+
+		type DecorateIn struct {
+			dig.In
+
+			Values []string `group:"values"`
+		}
+
+		type DecorateOut struct {
+			dig.Out
+
+			Values []string `group:"decoratedVals"`
+		}
+
+		type InvokeIn struct {
+			dig.In
+
+			Values []string `group:"decoratedVals"`
+		}
+
+		parent := dig.New()
+
+		require.NoError(t, parent.Provide(func() string { return "dog" }, dig.Group("values")))
+		require.NoError(t, parent.Provide(func() string { return "cat" }, dig.Group("values")))
+
+		child := parent.Scope("child")
+
+		parent.Decorate(func(i DecorateIn) DecorateOut {
+			var result []string
+			for _, val := range i.Values {
+				result = append(result, "happy "+val)
+			}
+			return DecorateOut{
+				Values: result,
+			}
+		})
+
+		child.Decorate(func(i DecorateIn) DecorateOut {
+			var result []string
+			for _, val := range i.Values {
+				result = append(result, "good "+val)
+			}
+			return DecorateOut{
+				Values: result,
+			}
+		})
+
+		child.Invoke(func(i InvokeIn) {
+			assert.ElementsMatch(t, []string{"happy dog", "happy cat", "good dog", "good cat"}, i.Values)
+		})
```

Comment:
Each level of decorate on a value group is an append, not a replace?
That's surprising behavior, IMO.
Should we do a replace, and if the user wants to append, they can copy items from the original slice too?

(Also, I guess for the other value group case above, we should also use ElementsMatch because this was not obvious with just the Contains check.)

Reply:
Fixed!